File: software/Navigator/path_optimization/path_pruning.py
import numpy as np

from bresenham3d import Bresenham3D

import numpy as np
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt

class PathPruning:

    def __init__(self, obstacle_distance=10):

        self.obstacle_distance = obstacle_distance

    '''
    for a path where there are multiple points that are collinear
    keep the first and last points and remove the rest.
    '''
    def remove_collinear_points(self, original_path):

        original_path = self.Remove(original_path)
        new_path = []
        length = len(original_path)
        new_path.append(original_path[0])

        for i in range(length)[2:]:

            distance13 = self.distance(original_path[i], original_path[i-2])
            distance12 = self.distance(original_path[i-1], original_path[i-2])
            distance23 = self.distance(original_path[i], original_path[i-1])

            if abs(distance13 - distance12 - distance23) < 0.001:
                continue

            else:
                new_path.append(original_path[i-1])

        new_path.append(original_path[-1])

        return new_path

    '''
    1. given a path: [a, b, c, d];
    2. iterate points in a way like [a, b], [a, c], [a, d], [b, c] ...
    3. draw a line with bresenham and see if obstacle point is in it
    4. if not, remove middle point and continue
    '''
    def path_pruning_bresenham3d(self, path, local_obstacle):

        final_path = []
        start_position = path[0]
        final_path.append(start_position)
        last_point = start_position

        for point in path:

            bresenham_path = Bresenham3D(start_position, point)

            for ob in local_obstacle:
                for path_point in bresenham_path:
                    if self.distance(ob, path_point) < self.obstacle_distance:

                        #print("ob in local path: ", ob)
                        final_path.append(last_point)
                        final_path.append(point)
                        start_position = point

            last_point = point

        final_path.append(path[-1])
        final_path = self.Remove(final_path)

        return final_path


    '''
    path smoothing using B spline
    '''
    def BSplinePathSmoothing(self, path):
        pass


    '''
    helper functions
    '''
    def Remove(self, duplicate):
        final_list = []
        for num in duplicate:
            if num not in final_list:
                final_list.append(num)
        return final_list


    def distance(self, p1, p2):
        x_distance = (p2[0] - p1[0])**2
        y_distance = (p2[1] - p1[1])**2
        z_distance = (p2[2] - p1[2])**2

        return np.sqrt(x_distance + y_distance + z_distance)



if __name__ == '__main__':

    def draw_path(path, ax, color='r', z_offset=0):
        x = []
        y = []
        z = []
        for point in path:
            x.append(point[0])
            y.append(point[1])
            z.append(point[2] + z_offset)

        ax.scatter(x, y, z, c=color, s=100)
        ax.plot(x, y, z, color=color)

    # generate a flight path
    path = []
    for i in range(10):
        path.append((i, 0, 0))

    for i in range(10):
        path.append((9, i, 0))

    for i in range(5):
        path.append((9, 9, i))

    # generate a obstacle
    obstacle = []
    for i in range(1, 5)[2:]:
        for j in range(1, 5)[2:]:
            for k in range(1, 16)[6:]:
                obstacle.append((i, j, k))


    ax = plt.gca(projection="3d")
    PathPruning = PathPruning(7.1)

    draw_path(obstacle, ax, 'r')
    draw_path(path, ax, 'r')

    # 1. remove collinear points
    processed_path_1 = PathPruning.remove_collinear_points(path)

    print("processed_path_1: ", processed_path_1)

    draw_path(processed_path_1, ax, 'g', 5)

    # 2. use raytracing to remove extra points
    processed_path_2 = PathPruning.path_pruning_bresenham3d(path, obstacle)

    print("processed_path_2: ", processed_path_2)

    draw_path(processed_path_2, ax, 'b', 10)

    plt.show()

File: software/ros_astar_C++/my_dir/path_optimization/path_pruning.cpp
#include <iostream>
#include <set> 
#include <vector>
#include <math.h>
#include<bits/stdc++.h>
#include <algorithm>

#include "bresenham3d.h"
#include "path_pruning.h"

using namespace std;




PathPruning::PathPruning(double obstacle_dist)
{
    obstacle_distance = obstacle_dist;
    //cout<<"In the constructor"<<obstacle_distance<<endl;
}

vector<Pair> PathPruning::remove_collinear_points(vector<Pair> original_path)
{
    double distance13, distance12, distance23;
    original_path = Remove(original_path);
    
    int length = original_path.size();
    new_path.push_back(original_path[0]);

    for(int i=2; i<length; i++)
    {
        distance13 = distance(original_path[i], original_path[i-2]);
        distance12 = distance(original_path[i-1], original_path[i-2]);
        distance23 = distance(original_path[i], original_path[i-1]);

        if(abs(distance13 - distance12 - distance23) < 0.001)
        {
            continue;
        }
        else
        {
            new_path.push_back(original_path[i-1]);
        }
    }

    new_path.push_back(original_path[original_path.size()-1]);
    return new_path; 
}

vector<Pair> PathPruning::Bresenham3D(Pair p1, Pair p2)
{
//print_pair(p1);
//print_pair(p2);
double x1 = p1.first;
double y1 = p1.second.first;
double z1 = p1.second.second;

double x2 = p2.first;
double y2 = p2.second.first;
double z2 = p2.second.second;

vector<Pair> ListOfPoints;
ListOfPoints.push_back(p1);


double dx = abs(x2 - x1);
double dy = abs(y2 - y1);
double dz = abs(z2 - z1);
//cout<<dx<<" "<<dy<<" "<<dz<<endl;

int xs, ys, zs;

if(x2>x1)
    xs = 1;
else
    xs = -1;

if (y2 > y1)
    ys = 1;
else
    ys = -1;

if (z2 > z1)
    zs = 1;
else
    zs = -1;

double temp1;
double temp2;
// Driving axis is X-axis"
if(dx >= dy && dx >= dz)
{
    temp1 = 2*dy - dx;
    temp2 = 2*dz - dx;
    while(x1!=x2)
    {
        x1  = x1 + xs;
        if(temp1 >= 0)
        {
            y1 = y1 + ys;
            temp1 = temp1 - 2*dx;
        }
        if(temp2 >= 0)
        {
            z1 = z1 + zs;
            temp2 = temp2 - 2*dx;
        }
        temp1 = temp1 + 2*dy;
        temp2 = temp2 + 2*dz;
        ListOfPoints.push_back(make_pair(x1,make_pair(y1,z1)));
    }
}

//Driving axis is Y-axis"
else if (dy >= dx && dy >= dz)
{
    temp1 = 2*dx - dy;
    temp2 = 2*dz - dy;
    while(y1!=y2)
    {
        y1  = y1 + ys;
        if(temp1 >= 0)
        {
            x1 = x1 + xs;
            temp1 = temp1 - 2*dy;
        }
        if(temp2 >= 0)
        {
            z1 = z1 + zs;
            temp2 = temp2 - 2*dy;
        }
        temp1 = temp1 + 2*dx;
        temp2 = temp2 + 2*dz;
        ListOfPoints.push_back(make_pair(x1,make_pair(y1,z1)));
    } 
}

//Driving axis is Z-axis"
else
{
    temp1 = 2*dy - dz;
    temp2 = 2*dx - dz;
    while(z1!=z2)
    {
        z1  = z1 + zs;
        if(temp1 >= 0)
        {
            y1 = y1 + ys;
            temp1 = temp1 - 2*dz;
        }
        if(temp2 >= 0)
        {
            x1 = x1 + xs;
            temp2 = temp2 - 2*dz;
        }
        temp1 = temp1 + 2*dy;
        temp2 = temp2 + 2*dx;
        ListOfPoints.push_back(make_pair(x1,make_pair(y1,z1)));
    }
}

return ListOfPoints;
}


vector<Pair> PathPruning::path_pruning_bresenham3d(vector<Pair> path, set<Pair> local_obstacle)
{
    Pair start_position = path[0];
    final_path.push_back(start_position);
    Pair last_point = start_position;
    vector<Pair> bresenham_path;
    //cout<<"In the funcion"<<obstacle_distance<<endl;

    for(int i=0; i<path.size(); i++)
    {
        bresenham_path = Bresenham3D(start_position, path[i]);
        //int count = 0;
        
        for (set<Pair>::iterator it=local_obstacle.begin(); it!=local_obstacle.end(); ++it)
        {
            for(int j=0; j<bresenham_path.size(); j++)
            {
                //if((*it).first == 4 && (*it).second.first == 3 && (*it).second.second == 7 && bresenham_path[j].first == 4 && bresenham_path[j].second.first == 2 && bresenham_path[j].second.second == 0)
                //{
                    //  cout<<distance(*it, bresenham_path[j])<<" "<<obstacle_distance<<endl;
                //} 
                //cout<<distance(*it, bresenham_path[j])<<" "<<obstacle_distance<<endl;
                if(distance(*it, bresenham_path[j]) < obstacle_distance)
                {
                    //count++;
                    final_path.push_back(last_point);
                    final_path.push_back(path[i]);
                    start_position = path[i];
                    //cout<<count<<endl;
                }
            }
        }
        last_point = path[i];
    }
    final_path.push_back(path[path.size()-1]);
    return final_path;
}
/*
vector<Pair> BSplinePathSmoothing(vector<Pair> path)
{
    return;
}*/

vector<Pair> PathPruning::Remove(vector<Pair> v)
{
vector<Pair>::iterator itr = v.begin();
set<Pair> s;

for (auto curr = v.begin(); curr != v.end(); ++curr) 
{
    if (s.insert(*curr).second)
        *itr++ = *curr;
}

v.erase(itr, v.end());
return v;
}

double PathPruning::distance(Pair pt1, Pair pt2)
{
    return(sqrt(pow((pt1.first - pt2.first),2)+pow((pt1.second.first - pt2.second.first),2)+pow((pt1.second.second - pt2.second.second),2)));
}

void PathPruning::print_pair(Pair p1)
{
    cout<<p1.first<<" "<<p1.second.first<<" "<<p1.second.second<<endl;
}

/*
int main()
{
    vector<Pair> path;
    for(int i=0; i<10; i++)
    {
        path.push_back(make_pair(i,make_pair(0,0)));
    }
    for(int i=0; i<10; i++)
    {
        path.push_back(make_pair(9,make_pair(i,0)));
    }
    for(int i=0; i<5; i++)
    {
        path.push_back(make_pair(9,make_pair(9,i)));
    }

    set<Pair> obstacle;
    for(int i=3; i<5; i++)
    {
        for(int j=3; j<5; j++)
        {
            for(int k=7;k<16;k++)
            {
                obstacle.insert(make_pair(i,make_pair(j,k)));
            }
        }
    }

    PathPruning p(7.1);

    //for (set<Pair>::iterator it=obstacle.begin(); it!=obstacle.end(); ++it)
    //{
    //    cout<<(*it).first<<" "<<(*it).second.first<<" "<<(*it).second.second<<endl;
    //}
    
     vector<Pair> a = p.remove_collinear_points(path);
    cout<<"a path"<<endl;
    for(int i=0; i<a.size(); i++)
    {
        cout<<a[i].first<<" "<<a[i].second.first<<" "<<a[i].second.second<<endl;
    }

    vector<Pair> b = p.path_pruning_bresenham3d(path, obstacle);
    cout<<"b path"<<endl;
    for(int i=0; i<b.size(); i++)
    {
        cout<<b[i].first<<" "<<b[i].second.first<<" "<<b[i].second.second<<endl;
    }


    return 0;
}
*/